Add documentation for failure stores. (#1368)
elastic/docs-content: manage-data/data-store/data-streams/failure-store.md

---
applies_to:
  stack: ga 8.19.0
  serverless: ga 9.1.0
---

# Failure store [failure-store]

> Failure stores are a secondary set of indices inside a data stream dedicated to storing failed documents. Failed documents are any documents that cause ingest pipeline exceptions or have a structure that conflicts with a data stream's mappings. These failures normally cause the indexing operation to fail, returning the error message in the response. 

Reviewer: ```suggestion
A failure store is a secondary set of indices inside a data stream, dedicated to storing failed documents. A failed document is any document that, without the failure store enabled, would cause an ingest pipeline exception or that has a structure that conflicts with a data stream's mappings. In the absence of the failure store, a failed document would cause the indexing operation to fail, with an error message returned in the operation response.
```

I took a stab at rephrasing to incorporate Pete's comment above, and also to introduce the terms "failure store" and "failed document" as singular rather than plural, which is our usual convention, but it's just a suggestion. I think the original works nicely too.

When a data stream's failure store is enabled, these failures are instead captured in a separate index and persisted to be analysed later. Clients receive a successful response with a flag indicating the failure was redirected. Failure stores do not capture failures caused by backpressure or document version conflicts. These failures are always returned as-is since they warrant specific action by the client.

## Set up a data stream failure store [set-up-failure-store]

> Each data stream has its own failure store that can be enabled to accept failures. By default, this failure store is disabled and any ingestion problems are raised in the response to write operations.

Reviewer: ```suggestion
Each data stream has its own failure store that can be enabled to accept failed documents. By default, this failure store is disabled and any ingestion problems are raised in the response to write operations.
```
Just for consistency with some of the other text.

### Set up for new data streams [set-up-failure-store-new]

> You can specify on a data stream's template if it should enable the failure store when it is first created. The `data_stream_options` field in a [template](../templates.md) contains the settings required to enable a data stream's failure store.

Reviewer: ```suggestion
You can specify in a data stream's [template](../templates.md) if it should enable the failure store when it is first created.
```
I'd put the link on the first instance of "template" and then we can remove the second sentence, since I think it's pretty similar to what you have above the example.

:::{note}
Unlike the `settings` and `mappings` fields on an [index template](../templates.md) which are repeatedly applied to new data stream write indices on rollover, the `data_stream_options` section of a template is applied to a data stream only once when the data stream is first created. To configure existing data streams, use the put [data stream options API](./failure-store.md).
:::

To enable the failure store on a new data stream, enable it in the `data_stream_options` of the template:

```console
PUT _index_template/my-index-template
{
  "index_patterns": ["my-datastream-*"],
  "data_stream": { },
  "template": {
    "data_stream_options": { <1>
      "failure_store": {
        "enabled": true <2>
      }
    }
  }
}
```

1. The options for a data stream to be applied at creation time.
2. The failure store feature will be enabled for new data streams that match this template.

After a matching data stream is created, its failure store will be enabled.

### Set up for existing data streams [set-up-failure-store-existing]

> Enabling the failure store via [index templates](../templates.md) can only affect data streams that are newly created. Existing data streams that use a template will not apply any changes to the template's `data_stream_options` after they have been created.

Reviewer: ```suggestion
Enabling the failure store via [index templates](../templates.md) can only affect data streams that are newly created. Existing data streams that use a template are not affected by changes to the template's `data_stream_options` field.
```
Just a tweak. :-)

To modify an existing data stream's options, use the [put data stream options](./failure-store.md) API:

```console
PUT _data_stream/my-datastream-existing/_options
{
  "failure_store": {
    "enabled": true <1>
  }
}
```

1. The failure store option will now be enabled.

> The failure store redirection can be disabled using this API as well. When the failure store is deactivated, only failed document redirection is halted. Any existing failure data in the data stream will remain until removed by manual deletion or by retention.

Reviewer: ```suggestion
The failure store redirection can be disabled using this API as well. When the failure store is deactivated, only failed document redirection is halted. Any existing failure data in the data stream will remain until removed by manual deletion or by an expired data retention setting.
```
Just a tweak, but please ignore if I've got it wrong.

Reviewer: Looks good!

If the document could have been redirected to a data stream's failure store but the failure store was disabled, then the `failure_store` field on the response will be `not_enabled`, and the response will display the error encountered as normal.

```console-result
{
  "error": {
    "root_cause": [ <1>
      {
        "type": "document_parsing_exception",
        "reason": "[1:53] failed to parse field [id] of type [long] in document with id 'Y0vQipYB_ZAKuDfZR4sR'. Preview of field's value: 'invalid_text'"
      }
    ],
    "type": "document_parsing_exception",
    "reason": "[1:53] failed to parse field [id] of type [long] in document with id 'Y0vQipYB_ZAKuDfZR4sR'. Preview of field's value: 'invalid_text'",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "For input string: \"invalid_text\""
    },
    "failure_store": "not_enabled" <2>
  },
  "status": 400 <3>
}
```

1. The failure is returned to the client as normal when the failure store is not enabled.
2. The response is annotated with a flag indicating the failure store could have accepted the document, but it was not enabled.

> 3. Status of 400 Bad Request due to the mapping problem.

Reviewer: ```suggestion
3. The response status is `400 Bad Request` due to the mapping problem.
```
(just for consistency with the previous example)

          {
            "type": "cluster_block_exception", <2>
            "reason": "index [.fs-my-datastream-2025.05.01-000002] blocked by: [FORBIDDEN/5/index read-only (api)];"
          }
        ]
      }
    ],
    "type": "document_parsing_exception", <3>
    "reason": "[1:53] failed to parse field [id] of type [long] in document with id 'Y0vQipYB_ZAKuDfZR4sR'. Preview of field's value: 'invalid_text'",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "For input string: \"invalid_text\""
    },
    "suppressed": [
      {
        "type": "cluster_block_exception",
        "reason": "index [.fs-my-datastream-2025.05.01-000002] blocked by: [FORBIDDEN/5/index read-only (api)];"
      }
    ],
    "failure_store": "failed" <4>
  },
  "status": 400 <5>
}
```

1. The root cause of the problem was a mapping mismatch.
2. The document could not be redirected because the failure store was not able to accept writes at this time due to an unforeseeable issue.
3. The complete exception tree is present on the response.
4. The response is annotated with a flag indicating the failure store would have accepted the document, but it was not able to.

> 5. Status of 400 Bad Request due to the original mapping problem.

Reviewer: ```suggestion
5. The response status is `400 Bad Request` due to the original mapping problem.
```

    
    `document.id`
    :   (`keyword`) The id of the original document at the time of failure.
    
    `document.routing`
    :   (`keyword`, optional) The routing of the original document at the time of failure if it was specified.
    
    `document.index`
    :   (`keyword`) The index that the document was being written to when it failed.

    `document.source`
    :   (unmapped object) The body of the original document. This field is unmapped and only present in the failure document's source. This prevents mapping conflicts in the failure store when redirecting failed documents. If you need to include fields from the original document's source in your queries, use [runtime fields](./failure-store.md) on the search request.

`error`
:   (`object`) Information about the failure that prevented this document from being indexed.

    `error.message`
    :   (`match_only_text`) The error message that describes the failure.

    `error.stack_trace`
    :   (`text`) A compressed stack trace from {{es}} for the failure.

    `error.type`

>     :   (`keyword`) The type classification of failure. Values are the same type returned within failed indexing API responses.

Reviewer: ```suggestion
    :   (`keyword`) The type classification of the failure. Values are the same type returned within failed indexing API responses.
```

            "index": "my-datastream-ingest",
            "source": {
              "random": 42
            }
          },
          "error": {
            "type": "illegal_argument_exception",
            "message": "field [data] not present as part of path [data]", <2>
            "stack_trace": """j.l.IllegalArgumentException: field [data] not present as part of path [data]
	at o.e.i.IngestDocument.getFieldValue(IngestDocument.java:202)
	at o.e.i.c.SetProcessor.execute(SetProcessor.java:86)
	... 14 more
""",
            "pipeline_trace": [
              "my-datastream-example-pipeline"
            ],
            "pipeline": "my-datastream-example-pipeline",
            "processor_type": "set"
          }
        }
      }
    ]
  }
}
```
1. The `document` field shows the state of the document is from before any pipeline executions.
2. The pipeline failed after the timestamp would have been added.

We can see that the document failed on the second processor in the pipeline. The first processor would have added a `@timestamp` field. Since the pipeline failed, we find that it has no `@timestamp` field added because it did not save any changes from before the pipeline failed.

> The second place failures can occur is during indexing. After the documents have been processed by any applicable pipelines, they are parsed using the index mappings before being indexed into the shard. If a document is sent to the failure store due to a failure in this process, then it will be stored as it was after any ingestion had occurred. This is because the original document is overwritten by the ingest pipeline changes by this point. This has the benefit of being able to see what the document looked like during the mapping and indexing phase of the write operation.

Reviewer: ```suggestion
The second time when failures can occur is during indexing. After the documents have been processed by any applicable pipelines, they are parsed using the index mappings before being indexed into the shard. If a document is sent to the failure store due to a failure in this process, then it will be stored as it was after any ingestion had occurred. This is because, by this point, the original document has already been overwritten by the ingest pipeline changes. This has the benefit of allowing you to see what the document looked like during the mapping and indexing phase of the write operation.
```
Small tweak.

          }
        }
      }
    ]
  }
}
```

Without tags in place it would not be as clear where in the pipeline we encountered the problem. Tags provide a unique identifier for a processor that can be quickly referenced in case of an ingest failure.

### Alerting on failed ingestion [recipes-alerting]

Since failure stores can be searched just like a normal data stream, we can use them as inputs to [alerting rules](./failure-store.md) in Kibana. Here is a simple alerting example to trigger on more than ten failures in the last five minutes for a data stream:

:::::{stepper}

::::{step} Create a failure store data view
If you want to use KQL or Lucene query types, you should first create a data view for your failure store data.
If you plan to use {{esql}} or the Query DSL query types, this step is not required.

Navigate to the data view page in Kibana and add a new data view. Set the index pattern to your failure store using the selector syntax.

:::{image} /manage-data/images/elasticsearch-reference-management_failure_store_alerting_create_data_view.png
:alt: create a data view using the failure store syntax in the index name
:::
::::

::::{step} Create new rule

> Navigate to Management / Alerts and Insights / Rules. Create a new rule. Choose the Elasticsearch query option.

Reviewer: ```suggestion
Navigate to Management / Alerts and Insights / Rules. Create a new rule. Choose the {{es}} query option.
```

---
applies_to:
  stack: ga 8.19.0
  serverless: ga 9.1.0
---

# Failure store [failure-store]

> Failure stores are a secondary set of indices inside a data stream dedicated to storing failed documents. Failed documents are any documents that cause ingest pipeline exceptions or have a structure that conflicts with a data stream's mappings. These failures normally cause the indexing operation to fail, returning the error message in the response. When a data stream's failure store is enabled, these failures are instead captured and persisted to be analysed later, returning a successful response to the client in the meantime.

Author: Added in 28287fe4

## Set up a data stream failure store [set-up-failure-store]

Each data stream has its own failure store that can be enabled to accept failures. By default, this failure store is disabled and any ingestion problems are raised in the response to write operations.

### Set up for new data streams [set-up-failure-store-new]

You can specify on a data stream's template if it should enable the failure store when it is first created. The `data_stream_options` field in a [template](../templates.md) contains the settings required to enable a data stream's failure store.

:::{note}

> Unlike the `settings` and `mappings` fields on an [index template](../templates.md) which are repeatedly applied to new data stream write indices over time, the `data_stream_options` section of a template is applied to a data stream only once when the data stream is first created. To configure existing data streams, use the put data stream options API.

Author: Added placeholder link in f1cb727a

:::

To enable the failure store on a new data stream, enable it in the `data_stream_options` of the template:

```console
PUT _index_template/my-index-template
{
  "index_patterns": ["my-datastream-*"],
  "data_stream": { },
  "template": {
    "data_stream_options": { <1>
      "failure_store": {
        "enabled": true <2>
      }
    }
  }
}
```

1. The options for a data stream to be applied at creation time.
2. The failure store feature will be enabled for new data streams that match this template.

After a matching data stream is created, its failure store will be enabled.

### Set up for existing data streams [set-up-failure-store-existing]

Enabling the failure store via [index templates](../templates.md) can only affect data streams that are newly created. Existing data streams that use a template will not apply any changes to the template's `data_stream_options` after they have been created.

> To modify an existing data stream's options, use the [put data stream options](./failure-store.md) API:

Author: Just a WIP tag - This is just a placeholder link for me to find and replace on later. I need to spend some time finding the right links for anything that has `./failure-store.md` as the link target.

```console
PUT my-datastream-existing/_options
{
  "failure_store": {
    "enabled": true <1>
  }
}
```

1. The failure store option will now be enabled.

The failure store redirection can be suspended using this API as well. When the failure store is disabled, only failed document redirection is halted. Any existing failure data in the data stream will remain until removed by deletion or by retention.

```console
PUT my-datastream-existing/_options
{
  "failure_store": {
    "enabled": false <1>
  }
}
```

1. Redirecting failed documents into the failure store will now be disabled.

### Enable failure store via cluster setting [set-up-failure-store-cluster-setting]

If you have a large number of existing data streams you may want an easier way to control if failures should be redirected. Instead of enabling the failure store using the [put data stream options](./failure-store.md) API, you can instead configure a set of patterns in the [cluster settings](./failure-store.md) which will enable the failure store feature by default.

> Configure a list of patterns using the `data_streams.failure_store.enabled` dynamic cluster setting. If a data stream matches a pattern in this setting and does not have the failure store explicitly disabled in its options, then the failure store will default to being enabled for that matching data stream.

Reviewer 2: I was wondering that as well when reading this.

> then the failure store will default to being enabled for that matching data stream

Makes it sounds like it's not applying to existing data streams, just act as a default.

> If you have a large number of existing data streams [...] you can instead configure a set of patterns in the cluster settings

Makes it sound the setting is an alternative to enabling the failure store one by one via the DS options API if you have a large number of existing data streams.

> Configure a list of patterns using the `data_streams.failure_store.enabled` dynamic cluster setting. If a data stream matches a pattern in this setting and does not have the failure store explicitly disabled in its options, then the failure store will default to being enabled for that matching data stream.

Author: Yeah, I tried to phrase this in a way that made it not seem like setting the property was the same as toggling the feature on permanently. Matching data streams only enable their failure stores so long as they are not explicitly disabled in the options, and only for as long as they match the setting. 

I've put up an edit that should simplify the explanation a bit. I also added an example of the explicit disabling of the failure store overriding the cluster setting.

```console
PUT _cluster/settings
{
  "persistent" : {
    "data_streams.failure_store.enabled" : [ "my-datastream-*", "logs-*" ] <1>
  }
}
```

1. Indices that match `my-datastream-*` or `logs-*` will redirect failures to the failure store unless explicitly disabled.

## Using a failure store [use-failure-store]

The failure store is meant to ease the burden of detecting and handling failures when ingesting data to {{es}}. Clients are less likely to encounter unrecoverable failures when writing documents, and developers are more easily able to troubleshoot faulty pipelines and mappings.

### Failure redirection [use-failure-store-redirect]

Once a failure store is enabled for a data stream it will begin redirecting documents that fail due to common ingestion problems instead of returning errors in write operations. Clients are notified in a non-intrusive way when a document is redirected to the failure store.

> Each data stream's failure store is made up of a list of indices that are dedicated to storing failed documents. These indices function much like a data stream's normal backing indices: There is a write index that accepts failed documents, they can be rolled over, and are automatically cleaned up over time subject to a lifecycle policy.

Author: Updated in 08b2ac31

	at o.e.x.s.AbstractXContentParser.longValue(AbstractXContentParser.java:210)
	... 26 more
|document_parsing_exception
```

:::{note}
Because the `document.source` field is unmapped, it is absent from the SQL results.
:::
::::
:::::

Failure documents have a uniform structure that is handled internally by {{es}}.

`@timestamp`
:   (`date`) The timestamp at which the document encountered a failure in {{es}}.

`document`
:   (`object`) The document at time of failure. If the document failed in an ingest pipeline, then the document will be the unprocessed version of the document as it arrived in the original indexing request. If the document failed due to a mapping issue, then the document will be as it was after any ingest pipelines were applied to it.
    
    `document.id`
    :   (`keyword`) The id of the original document at the time of failure.
    
    `document.routing`
    :   (`keyword`, optional) The routing of the original document at the time of failure if it was specified.
    
    `document.index`
    :   (`keyword`) The index that the document was being written to when it failed.

    `document.source`

>     :   (unmapped object) The body of the document. This field is unmapped and unindexed to ensure failures are indexed reliably. If you need to include fields from the document source in your queries, use [runtime fields](./failure-store.md) on the search request.

Reviewer 2: > unindexed to ensure failures are indexed reliably.

I think I know what you mean but this sounds a bit confusing.

Author: Reworded a bit in 1591e7bb

      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false,
      "rollover_on_write": false,
      "index_mode": "standard",
      "failure_store": { <1>
        "enabled": true,
        "rollover_on_write": false,
        "indices": [
          {
            "index_name": ".fs-my-datastream-2025.05.05-000002",
            "index_uuid": "oYS2WsjkSKmdazWuS4RP9Q",
            "managed_by": "Data stream lifecycle"  <2>
          }
        ],
        "lifecycle": {
          "enabled": true,
          "effective_retention": "30d",  <3> 
          "retention_determined_by": "default_failures_retention"  <4>
        }
      }
    }
  ]
}
```
1. Information about the failure store is presented in the response under its own field.
2. Indices are managed by data stream lifecycles by default.
3. An effective retention period of thirty days (30d) is present by default.
4. The retention is currently determined by the default.  

:::{note}
The default retention respects any maximum retention values. If [maximum retention](./failure-store.md) is configured lower than thirty days then the maximum retention will be used as the default value.
:::

> You can update the default retention period for failure stores in your deployment by updating the `data_streams.lifecycle.retention.failures_default` cluster setting. Data streams that have no retention configured on their failure stores will use this value to determine their retention period.

Reviewer 2: Will this change the retention for existing data streams?

Author: Yes, changed to "New and existing data streams that have no retention configured on their failure store will use this value..." in cec305f9

```console
PUT _cluster/settings
{
  "persistent": {
    "data_streams.lifecycle.retention.failures_default": "15d"
  }
}
```

You can also specify the failure store retention period for a data stream on its data stream options. These can be specified via the index template for new data streams, or via the [put data stream options](./failure-store.md) API for existing data streams.

```console
PUT _data_stream/my-datastream/_options
{
    "failure_store": {
        "enabled": true, <1>
        "lifecycle": {
            "data_retention": "10d" <2>
        }
    }
}
```
1. Ensure that the failure store remains enabled.
2. Set only this data stream's failure store retention to ten days.

### Add and remove from failure store [manage-failure-store-indices]

> Failure stores support adding and removing indices from them using the [modify data stream](./failure-store.md) API.

Reviewer 2: What happens if you add a failure store backing index that has incompatible mappings? Are we doing validations when adding a backing index or would it fail at runtime? Not suggesting one way is better than the other but maybe we should describe what happens here to set expectations.

Author: There's no special handling for mappings when adding an index to the failure store. You could add a completely unrelated index to the failure store and we allow it. Indices that are added to a data stream can never be treated as a write index, so we're less worried about their mappings than when doing a rollover operation. Even if the failure store is empty and we add a random index, the failure store is still marked for lazy rollover and will create a write index on redirection.

---
applies_to:
  stack: ga 8.19.0
  serverless: ga 9.1.0
---

# Failure store [failure-store]

> Failure stores are a secondary set of indices inside a data stream dedicated to storing failed documents. Failed documents are any documents that cause ingest pipeline exceptions or have a structure that conflicts with a data stream's mappings. These failures normally cause the indexing operation to fail, returning the error message in the response. 

Reviewer 3: Nit: Maybe replace "These failures normally cause" with something like "Without the failure store, these failures would". I don't think it's obvious from your phrasing what "normally" means — like, maybe you're explaining the behaviour _with_ the failure store, since this is a page about failure store? — and it's better to be explicit. (I am aware of this because I've confused people before by saying things like this!)

When a document bound for a data stream encounters a problem during its ingestion, the response is annotated with the `failure_store` field which describes how {{es}} responded to that problem. The `failure_store` field is present on both the [bulk](./failure-store.md) and [index](./failure-store.md) API responses when applicable. Clients can use this information to augment their behavior based on the response from {{es}}.

Here we have a bulk operation that sends two documents. Both are writing to the `id` field which is mapped as a `long` field type. The first document will be accepted, but the second document would cause a failure because the value `invalid_text` cannot be parsed as a `long`. This second document will be redirected to the failure store: 

```console

> POST my-datastream/_bulk

Reviewer 4: Should be 

`POST my-datastream-new/_bulk` to match the data stream template at the top of the page.

---
applies_to:

>   stack: ga 8.19.0

Reviewer 5: Can we put both 8.19 and 9.1 here? I want to make sure a user doesn't assume 9.0 > 8.19 and therefore should have this feature (even though it doesn't)

>   serverless: ga 9.1.0

Reviewer 5: Does Serverless have versions for the docs?

---

# Failure store [failure-store]

A failure store is a secondary set of indices inside a data stream, dedicated to storing failed documents. A failed document is any document that, without the failure store enabled, would cause an ingest pipeline exception or that has a structure that conflicts with a data stream's mappings. In the absence of the failure store, a failed document would cause the indexing operation to fail, with an error message returned in the operation response.

> When a data stream's failure store is enabled, these failures are instead captured in a separate index and persisted to be analysed later. Clients receive a successful response with a flag indicating the failure was redirected. Failure stores do not capture failures caused by backpressure or document version conflicts. These failures are always returned as-is since they warrant specific action by the client.

Reviewer 5: I think it would be worth separating out the `Failure stores do not capture failures caused by backpressure or document version conflicts. These failures are always returned as-is since they warrant specific action by the client.` section into a "note" or "important" block, so it doesn't get missed.

## Set up a data stream failure store [set-up-failure-store]

Each data stream has its own failure store that can be enabled to accept failed documents. By default, this failure store is disabled and any ingestion problems are raised in the response to write operations.

### Set up for new data streams [set-up-failure-store-new]

You can specify in a data stream's [index template](../templates.md) if it should enable the failure store when it is first created.

:::{note}

> Unlike the `settings` and `mappings` fields on an [index template](../templates.md) which are repeatedly applied to new data stream write indices on rollover, the `data_stream_options` section of a template is applied to a data stream only once when the data stream is first created. To configure existing data streams, use the put [data stream options API](indices-put-data-stream-options).

Reviewer 5: Will the `indices-put-data-stream-options` link work? Or do we need to fully-qualify it with the URL?

> Unlike the `settings` and `mappings` fields on an [index template](../templates.md) which are repeatedly applied to new data stream write indices on rollover, the `data_stream_options` section of a template is applied to a data stream only once when the data stream is first created. To configure existing data streams, use the put [data stream options API](indices-put-data-stream-options).

Reviewer 6: ```suggestion
Unlike the `settings` and `mappings` fields on an [index template](../templates.md) which are repeatedly applied to new data stream write indices on rollover, the `data_stream_options` section of a template is applied to a data stream only once when the data stream is first created. To configure existing data streams, use the put [data stream options API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-put-data-stream-options).
```

Reviewer 6: wonder if we need links to the serverless versions of these APIs too
https://www.elastic.co/docs/api/doc/elasticsearch-serverless/operation/operation-indices-put-data-stream-options

### Enable failure store via cluster setting [set-up-failure-store-cluster-setting]

If you have a large number of existing data streams you may want to enable their failure stores in one place. Instead of updating each of their options individually, set `data_streams.failure_store.enabled` to a list of index patterns in the [cluster settings](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-put-settings). Any data streams that match one of these patterns will operate with their failure store enabled.

```console
PUT _cluster/settings
{
  "persistent" : {
    "data_streams.failure_store.enabled" : [ "my-datastream-*", "logs-*" ] <1>
  }
}
```
1. Indices that match `my-datastream-*` or `logs-*` will redirect failures to the failure store unless explicitly disabled.

Matching data streams will ignore this configuration if the failure store is explicitly enabled or disabled in their [data stream options](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-put-data-stream-options).

```console
PUT _cluster/settings
{
  "persistent" : {
    "data_streams.failure_store.enabled" : [ "my-datastream-*", "logs-*" ] <1>
  }
}
```

> ```console

Reviewer 5: I think these two lines need to be removed to make this one continuous console block, since we use `<2>` below

          }
        ]
      }
    ],
    "type": "document_parsing_exception", <3>
    "reason": "[1:53] failed to parse field [id] of type [long] in document with id 'Y0vQipYB_ZAKuDfZR4sR'. Preview of field's value: 'invalid_text'",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "For input string: \"invalid_text\""
    },
    "suppressed": [
      {
        "type": "cluster_block_exception",
        "reason": "index [.fs-my-datastream-2025.05.01-000002] blocked by: [FORBIDDEN/5/index read-only (api)];"
      }
    ],
    "failure_store": "failed" <4>
  },
  "status": 400 <5>
}
```

1. The root cause of the problem was a mapping mismatch.
2. The document could not be redirected because the failure store was not able to accept writes at this time due to an unforeseeable issue.
3. The complete exception tree is present on the response.
4. The response is annotated with a flag indicating the failure store would have accepted the document, but it was not able to.
5. The response status is `400 Bad Request` due to the original mapping problem.

### Searching failures [use-failure-store-searching]

> Once you have accumulated some failures, they can be searched much like a regular index.

Reviewer 5: ```suggestion
Once you have accumulated some failures, the failure store can be searched much like a regular data stream.
```

:::{warning}
Documents redirected to the failure store in the event of a failed ingest pipeline will be stored in their original, unprocessed form. If an ingest pipeline normally redacts sensitive information from a document, then failed documents in their original, unprocessed form may contain sensitive information.

> Furthermore, failed documents are likely to be structured differently than normal data in a data stream, and thus are not supported by [document level security](../../../deploy-manage/users-roles/cluster-or-deployment-auth/controlling-access-at-document-field-level.md#document-level-security) or [field level security](../../../deploy-manage/users-roles/cluster-or-deployment-auth/controlling-access-at-document-field-level.md#field-level-security).

Reviewer 7: This is actually not true. We decided to support FLS and DLS against failure store. The main reason is because it would not be possible to prevent using FLS/DLS when users define implicit `read` access to backing `.fs*` indices. It felt wrong to prevent using FLS/DLS when users defined explicit `read_failure_store` access to data streams with FLS/DLS restrictions. Our biggest concern was that users would be expecting the DLS/FLS to stop certain docs/fields from being visible and it wouldn't. 

Right now, if users include FLS/DLS when granting access to failure store then we'll try to honour it. We should just make sure to highlight it (which you already did) that these documents are structured differently, and because they may contain sensitive information, the users should take extra care when defining access to them.

Furthermore, failed documents are likely to be structured differently than normal data in a data stream, and thus special care should be taken when making use of [document level security](../../../deploy-manage/users-roles/cluster-or-deployment-auth/controlling-access-at-document-field-level.md#document-level-security) or [field level security](../../../deploy-manage/users-roles/cluster-or-deployment-auth/controlling-access-at-document-field-level.md#field-level-security). Any security policies that expect to utilize these features for both regular documents and failure documents should account for any differences in document structure between the two document types.

To limit visibility on potentially sensitive data, users require the [`read_failure_store`](../../../deploy-manage/users-roles/cluster-or-deployment-auth/elasticsearch-privileges.md#privileges-list-indices) index privilege for a data stream in order to search that data stream's failure store data.
:::

Searching a data stream's failure store can be done by making use of the existing search APIs available in {{es}}. 

> To indicate that the search should be performed on failure store data, use the [index component selector syntax](elasticsearch://reference/elasticsearch/rest-apis/api-conventions.md#api-component-selectors) to indicate which part of the data stream to target in the search operation. Appending the `::failures` suffix to the name of the data stream indicates that the operation should be performed against that data stream's failure store instead of its regular backing indices.

Reviewer 6: ```suggestion
To indicate that the search should be performed on failure store data, use the [index component selector syntax](elasticsearch://reference/elasticsearch/rest-apis/api-conventions.md) to indicate which part of the data stream to target in the search operation. Appending the `::failures` suffix to the name of the data stream indicates that the operation should be performed against that data stream's failure store instead of its regular backing indices.
```
